pkg/etcdutil/etcdutil.go

```diff
@@ -292,5 +308,98 @@ func TakeAndSaveFullSnapshot(ctx context.Context, client client.MaintenanceClose
 	metrics.SnapshotDurationSeconds.With(prometheus.Labels{metrics.LabelKind: brtypes.SnapshotKindFull, metrics.LabelSucceeded: metrics.ValueSucceededTrue}).Observe(timeTaken.Seconds())
 	logger.Infof("Total time to save full snapshot: %f seconds.", timeTaken.Seconds())
 
+	defer func() {
+		if err := os.Remove(snapshotTempDBPath); err != nil {
+			logger.Warnf("failed to remove temporary full snapshot file: %v", err)
+		}
+		if err := snapshotData.Close(); err != nil {
+			logger.Warnf("failed to close snapshot data file: %v", err)
+		}
+	}()
```

Comment:
The `defer func()` as it is at line 311 might not ever get invoked causing resource leakage in case an error occurs anywhere before line 311. So it would be better to break it up into multiple defer calls and invoke as soon as possible.

The defer of `os.Remove()` should be at line 283, since it is established that the file is successfully created after the error check before that.

The defer of `snapshotData.Close()` should be at line 294 since we are writing to `snapshotData` in line 287, and we need to check for an error with the call at that line to ensure that `snapshotData` is not a `nil` and hasn't errored.

Reply:
address your review comments. PTAL